### ISSUE TITLE
Error linking test-wasi-import-object in runtime-c-api tests

### ISSUE THREAD

**Issue by AdamSLevy** [wasmerio/wasmer#988]
### Describe the bug

I am attempting to run the tests in `runtime-c-api` but the `make` step fails during `cargo test` due to `tests/test-wasi-import-object` being removed during some prior step.

### Steps to reproduce

On the latest commit on master, `6831a6d8`

```
$ cd lib/runtime-c-api
$ cargo test -- --nocapture
...
Running command: `make` args: ["-Wdev", "-Werror=dev"]
output:
status: 2
stdout:
Scanning dependencies of target test-module-import-instantiate
[  2%] Building C object CMakeFiles/test-module-import-instantiate.dir/test-module-import-instantiate.c.o
[  5%] Linking C executable test-module-import-instantiate
[  5%] Built target test-module-import-instantiate
Scanning dependencies of target test-context
[  8%] Building C object CMakeFiles/test-context.dir/test-context.c.o
[ 11%] Linking C executable test-context
[ 11%] Built target test-context
Scanning dependencies of target test-module-serialize
[ 14%] Building C object CMakeFiles/test-module-serialize.dir/test-module-serialize.c.o
[ 17%] Linking C executable test-module-serialize
[ 17%] Built target test-module-serialize
Scanning dependencies of target test-module-imports
[ 20%] Building C object CMakeFiles/test-module-imports.dir/test-module-imports.c.o
[ 23%] Linking C executable test-module-imports
[ 23%] Built target test-module-imports
Scanning dependencies of target test-import-function
[ 26%] Building C object CMakeFiles/test-import-function.dir/test-import-function.c.o
[ 29%] Linking C executable test-import-function
[ 29%] Built target test-import-function
Scanning dependencies of target test-instantiate
[ 32%] Building C object CMakeFiles/test-instantiate.dir/test-instantiate.c.o
[ 35%] Linking C executable test-instantiate
[ 35%] Built target test-instantiate
Scanning dependencies of target test-wasi-import-object
[ 38%] Building C object CMakeFiles/test-wasi-import-object.dir/test-wasi-import-object.c.o
[ 41%] Linking C executable test-wasi-import-object

stderr:
/usr/bin/ld: CMakeFiles/test-wasi-import-object.dir/test-wasi-import-object.c.o: in function `main':
test-wasi-import-object.c:(.text+0x63e): undefined reference to `wasmer_wasi_generate_import_object'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/test-wasi-import-object.dir/build.make:85: test-wasi-import-object] Error 1
make[1]: *** [CMakeFiles/Makefile2:284: CMakeFiles/test-wasi-import-object.dir/all] Error 2
make: *** [Makefile:95: all] Error 2

thread 'test_c_api' panicked at 'Command failed with exit status: ExitStatus(ExitStatus(512))', lib/runtime-c-api/tests/runtime_c_api_tests.rs:41:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
test test_c_api ... FAILED

failures:

failures:
    test_c_api

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--test runtime_c_api_tests'
```
Running `git status` shows that the `tests/test-wasi-import-object` has been removed.

```
$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    tests/test-wasi-import-object

no changes added to commit (use "git add" and/or "git commit -a")
```

**Comment by MarkMcCaskey** [comment]
Thanks for reporting this!

**Comment by MarkMcCaskey** [comment]
I believe I've made a fix in #991 ! It works for me, let me know if you run into any issues with it!

**Comment by AdamSLevy** [comment]
I am still getting errors: https://gist.github.com/ba9c925f9f1094481bca3af4fc800db9

**Comment by MarkMcCaskey** [comment]
Oh, to clarify it assumes default features.  You compiled the C API without default features, right?  If so, you'll need to run something like `cargo test --no-default-features --features "cranelift-backend"` for this to work

**Comment by AdamSLevy** [comment]
For this test I did exactly this:

```
git fetch
git checkout fix/c-api-tests-wasi
cd lib/runtime-c-api
cargo build
cargo test
```
and I got the output that I posted above.

When I ran `cargo test --no-default-features --features "cranelift-backend"` the tests passed. 

So I'm a little confused by what you said. I compiled it with default features but the test failed. When I override the defaults, the tests passed. Can you explain the defaults and what the test is complaining about when I use the defaults?

**Comment by MarkMcCaskey** [comment]
Sure!  So by default the c-api compiles with WASI support, the error you're getting is a linker error: the C code which is a test in this case is trying to use WASI specific functions but they're not in the library.

Notably, from looking at the `CMakeLists.txt`, it looks like we use the release form of the library, not the `debug` one, so `cargo build` in runtime-c-api won't do anything as far as the tests are concerned.  If you run `make capi` at the top level Makefile, it will run the correct command (which in this case is just `cargo build --release` in the runtime-c-api crate).

The fix I tried to implement checks the same features that are used to decide if the WASI-specific functions are compiled into the library.  Unfortunately these can be run at two different points in time so they can get out of sync.

If the C API is compiled in release mode it should work.  Though the tests must be compiled with the same features as the library.  This was the first thing that came to mind -- perhaps we want to generate the tests at build time and just run them when running `cargo test`, that should prevent this kind of thing from happening 🤔 .

In summary, I believe things are working correctly and `cargo build --release` will fix it.  If that's not the case, then there's likely a bug!

**Comment by AdamSLevy** [comment]
>  the C code which is a test in this case is trying to use WASI specific functions but they're not in the library.

Why are these functions only available in the release but not the debug version?

**Comment by AdamSLevy** [comment]
Also now that I built the api with --release the tests are passing.

**Comment by MarkMcCaskey** [comment]
Nice!  Glad that fixed it.

It is pretty confusing though, I've been bitten by not recompiling in release mode in the past, too, when working on the runtime-c-api.  I'm not sure how to fix this generally.

> Why are these functions only available in the release but not the debug version?

They're available in both, it's just that tests specifically only look for the release version of the library: https://github.com/wasmerio/wasmer/blob/master/lib/runtime-c-api/tests/CMakeLists.txt#L24

**Comment by MarkMcCaskey** [comment]
I think I figured out how this happened!  You probably compiled Wasmer in release mode before we had WASI in the C api.  So the tests were using an old c-api library.  That would explain how this could happen naturally.  Sorry about the inconvenience!
